src/lsp4clj/json_rpc.clj

```diff
@@ -0,0 +1,115 @@
+(ns lsp4clj.json-rpc
+  "Models LSP JSON-RPC as core.async channels of messages (Clojure hashmaps).
+
+  https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#baseProtocol"
+  (:require
+   [camel-snake-kebab.core :as csk]
+   [camel-snake-kebab.extras :as cske]
+   [cheshire.core :as json]
+   [clojure.core.async :as async]
+   [clojure.java.io :as io]
+   [clojure.string :as string]))
+
+(set! *warn-on-reflection* true)
+
+(defn ^:private read-n-chars [^java.io.InputStream input content-length charset-s]
+  (let [buffer (byte-array content-length)]
+    (loop [total-read 0]
+      (when (< total-read content-length)
+        (let [new-read (.read input buffer total-read (- content-length total-read))]
+          (when (< new-read 0)
+            ;; TODO: return nil instead?
+            (throw (java.io.EOFException.)))
+          (recur (+ total-read new-read)))))
+    (String. ^bytes buffer ^String charset-s)))
+
+(defn ^:private parse-header [line headers]
+  (let [[h v] (string/split line #":\s*" 2)]
+    (when-not (contains? #{"Content-Length" "Content-Type"} h)
+      (throw (ex-info "unexpected header" {:line line})))
+    (assoc headers h v)))
+
+(defn ^:private parse-charset [content-type]
+  (or (when content-type
+        (when-let [[_ charset] (re-find #"; charset=(.*)$" content-type)]
```

Comment:
I think it might be a good idea (now or eventually) to use a content-type parser here. 
At least, I would check charset in a case-insensitive way.

Reply:
Not a bad idea. I was going off lsp4j, which is [equally naive](https://github.com/eclipse/lsp4j/blob/5605ca82252833777dbef82ca13a403dbfb9d906/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/StreamMessageProducer.java#L161). I'm going to hold off on importing a real header parser, but I'll make the regex case-insensitive.